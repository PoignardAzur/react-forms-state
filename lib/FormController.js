'use babel'
// @flow

import React from 'react';
import { fromJS, is } from "immutable";

export default function FormController(
  applyControl = state => state,
  convertIn = (value, props) => value,
  convertOut = (value, props) => value
): () => React.Component<any, any, any> {
  return function(WrappedComponent: React.Component<any, any, any>): React.Component<any, any, any> {
    class Controller extends React.Component {
      constructor(props) {
        super();
        this.state = this._getInitialValue(props);
        this.rootDispatcherGetter = null;
      }

      componentWillReceiveProps(nextProps) {
        if (!is(fromJS(this._getInitialValue(nextProps)), fromJS(this._getInitialValue(this.props))))
          this.setState(this._getInitialValue(nextProps));
      }

      _getInitialValue(props) {
        let v = typeof props.initial != 'function' ? props.initial : props.initial(props);
        return convertIn(v, this.props);
      }

      _mergeValues() {
        return fromJS(this.state).mergeDeep(this.rootDispatcherGetter ? fromJS(this.rootDispatcherGetter()) : fromJS({}));
      }

      _submit() {
        let newValue = this._mergeValues();
        if (!is(newValue, fromJS(this._getInitialValue(this.props)))) {
          this.props.onSubmit(convertOut(newValue.toJS(), this.props));
        }
      }

      getChildContext() {
        return {
          attachToController: (getter) => {
            this.rootDispatcherGetter = getter;
          },
        };
      }

      render() {
        return (
          <WrappedComponent
            {...this.props}
            onChange={value => this.setState(applyControl(value, this.props))}
            value={this.state}
            onSubmit={() => this._submit()}
          />
        );
      }
    };

    Controller.childContextTypes = {
      attachToController: React.PropTypes.func,
    };

    return Controller;
  };
}
