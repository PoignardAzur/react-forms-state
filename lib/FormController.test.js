"use babel";
// @flow

import React from "react";
import { shallow, mount } from "enzyme";
import FormController from "./FormController";
import { fromJS } from "immutable";

describe("FormController", () => {
  it("should call onSubmit with the converted out form state on submit", () => {
    let Form = () => <div />;
    let ControlledForm = FormController(
      state => state,
      inValue => ({ name: inValue.firstname }),
      out => ({ firstname: out.name })
    )(Form);
    let spy = jest.fn();
    let subject = mount(
      <ControlledForm initial={{ firstname: "Alan" }} onSubmit={spy} />
    );

    subject.find(Form).at(0).prop("onChange")("Mr Joe", "name");
    subject.find(Form).at(0).prop("onSubmit")();

    expect(spy).toBeCalledWith({ firstname: "Mr Joe" });
  });

  it("should handle initial as an object", done => {
    let Form = () => <div />;
    let ControlledForm = FormController(
      state => state,
      inValue => ({ name: inValue.firstname }),
      out => ({ firstname: out.name })
    )(Form);
    let spy = jest.fn();
    let subject = mount(
      <ControlledForm
        initial={{ firstname: "Alan" }}
        onSubmit={spy}
        firstname="Alan"
      />
    );

    subject.find(Form).at(0).prop("valueChangeObs").subscribe({
      next: e => {
        expect(e.value.name).toBe("Alan");
        done();
      }
    });
  });

  it("should handle initial as a function", done => {
    let Form = () => <div />;
    let ControlledForm = FormController(
      state => state,
      inValue => ({ name: inValue.firstname }),
      out => ({ firstname: out.name })
    )(Form);
    let spy = jest.fn();
    let subject = mount(
      <ControlledForm
        initial={props => ({ firstname: props.firstname })}
        onSubmit={spy}
        firstname="Alan"
      />
    );

    subject.find(Form).at(0).prop("valueChangeObs").subscribe({
      next: e => {
        expect(e.value.name).toBe("Alan");
        done();
      }
    });
  });

  it("should refresh the state when initial change", done => {
    let Form = () => <div />;
    let ControlledForm = FormController(
      state => state,
      inValue => ({ name: inValue.firstname }),
      out => ({ firstname: out.name })
    )(Form);
    let subject = mount(<ControlledForm initial={{ firstname: "Alan" }} />);

    subject.find(Form).at(0).prop("valueChangeObs").skip(1).subscribe({
      next: e => {
        expect(e.value.name).toBe("Zanatta");
        done();
      }
    });

    subject.setProps({ initial: { firstname: "Zanatta" } });
  });

  it("should call applyControl on input value change before setting the new state", done => {
    let Form = () => <div />;
    let ControlledForm = FormController(
      state =>
        fromJS(state).update("name", name => name.replace(" ", "")).toJS(),
      inValue => ({ name: inValue.firstname }),
      out => ({ firstname: out.name })
    )(Form);
    let spy = jest.fn();
    let subject = mount(
      <ControlledForm initial={{ firstname: "Alan" }} onSubmit={spy} />
    );

    subject.find(Form).at(0).prop("valueChangeObs").skip(1).subscribe({
      next: e => {
        expect(e.value.name).toBe("MrJoe");
        done();
      }
    });

    subject.find(Form).at(0).prop("onChange")("Mr Joe", "name");
  });

  it("should call getUncontrolledState of child if it exists and merge states", () => {
    class Form extends React.Component {
      componentDidMount() {
        this.context.attachToController(() => ({ coucou: "salut" }));
      }
      render() {
        return <div />;
      }
    }
    Form.contextTypes = { attachToController: React.PropTypes.func };
    let ControlledForm = FormController(
      state =>
        fromJS(state).update("name", name => name.replace(" ", "")).toJS(),
      inValue => ({ name: inValue.firstname }),
      out => ({ firstname: out.name, coucou: out.coucou })
    )(Form);
    let submitSpy = jest.fn();
    let subject = mount(
      <ControlledForm initial={{ firstname: "Alan" }} onSubmit={submitSpy} />
    );

    subject.find(Form).at(0).prop("onSubmit")();

    expect(submitSpy.mock.calls[0][0]).toEqual({
      firstname: "Alan",
      coucou: "salut"
    });
  });
});
