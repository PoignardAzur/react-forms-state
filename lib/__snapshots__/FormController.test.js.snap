// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormController should send an event when state has changed 1`] = `
Object {
  "name": "Mr Joe",
}
`;
